### PR TITLE
feat: Added network and activity logs to Web Chat log panel

### LINKED ISSUE #6511: Support activity and bot state inspection in the embedded Web Chat control
- [x] Capture message / trace / event activities (P0)
- [x] Capture network activity (P0)
- [x] Show selected item in inspector (P0)
- [ ] Fix layout so that Web Chat panel **does not** overlap with inspector pane (P2)
- [x] Expose runtime messages in "Output" tab (P0)
- [ ]~~Add different inspector experience for LUIS traces (P1)~~
- [ ] Selected inspected state in log -- link log item to item shown in inspector pane (P1)
- [ ] On initial load, select most recent activity (P1)

### PR DESCRIPTION
**Pull request by tonyanziano**
## Description

This PR adds network and Web Chat activity logs to the Web Chat log panel.

## Task Item

Fixes #6511

image

image

### CODE REVIEW

File: Composer/packages/client/src/components/WebChat/WebChatPanel.tsx [hunk 1]
```diff
@@ -50,35 +54,50 @@ export const WebChatPanel: React.FC<WebChatPanelProps> = ({
...
-            }, 300);
+        // set up Web Chat traffic listener
+        webChatTrafficChannel.current = new WebSocket(`ws://localhost:${conversationServerPort}/ws/traffic`);
+        if (webChatTrafficChannel.current) {
+          webChatTrafficChannel.current.onmessage = (event) => {
+            const data:
+              | ConversationActivityTraffic
+              | ConversationNetworkTrafficItem
+              | ConversationNetworkErrorItem = JSON.parse(event.data);
+            if (data.trafficType === 'network') {
```
Comment (hatpick) [inline Composer/packages/client/src/components/WebChat/WebChatPanel.tsx:71]: nit: prefer switch/case, if not, use ```else if``` for the following conditions

File: Composer/packages/client/src/components/WebChat/WebChatPanel.tsx [hunk 2]
```diff
@@ -50,35 +54,50 @@ export const WebChatPanel: React.FC<WebChatPanelProps> = ({
...
+            }
+            if (data.trafficType === 'activity') {
+              appendWebChatTraffic(
+                projectId,
+                data.activities.map((a) => ({ activity: a, timestamp: a.timestamp, trafficType: data.trafficType }))
+              );
+            }
+            if (data.trafficType === 'networkError') {
+              appendWebChatTraffic(projectId, data);
+              setTimeout(() => {
```
Comment (hatpick) [inline Composer/packages/client/src/components/WebChat/WebChatPanel.tsx:82]: why the timeout?

File: Composer/packages/client/src/components/WebChat/WebChatPanel.tsx [hunk 3]
```diff
@@ -50,35 +54,50 @@ export const WebChatPanel: React.FC<WebChatPanelProps> = ({
...
-          status: response.status,
-          logType: 'Error',
-          message: formatMessage('An error occurred connecting initializing the DirectLine server'),
+        const err: ConversationNetworkErrorItem = {
+          error: {
+            message: formatMessage('An error occurred connecting initializing the DirectLine server'),
+          },
+          request: { route: 'conversations/ws/port', method: 'GET', payload: {} },
+          response: { payload: response.data, statusCode: response.status },
+          timestamp: new Date().toISOString(),
```
Comment (hatpick) [inline Composer/packages/client/src/components/WebChat/WebChatPanel.tsx:97]: do we want toISOString or locale string?

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  border: none;
+`;
+
+const pivotStyles = css`
+  margin-bottom: 8px;
+`;
+
+type WebChatInspectorPaneProps = {
+  inspectionData?: WebChatInspectionData;
+  setInspectionData: (data: WebChatInspectionData) => void;
```
Comment (hatpick) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx:27]: onUpdateInspectionData, without ```on``` prefix it sounds like an imperative method. Also ```set``` prefix is usually used for state mutators

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+export const WebChatInspectorPane: React.FC<WebChatInspectorPaneProps> = (props) => {
+  const { inspectionData, setInspectionData } = props;
+  const handleInspectorTabClick = useCallback(
+    (item: PivotItem | undefined) => {
+      if (item && inspectionData) {
+        if (item.props.itemKey === 'networkReq') {
+          setInspectionData({ ...inspectionData, mode: 'request' });
+        }
+        if (item.props.itemKey === 'networkRes') {
```
Comment (hatpick) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx:54]: else if

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+    default:
+      return type || 'Unknown';
+  }
+};
+
+export const WebChatInspectorPane: React.FC<WebChatInspectorPaneProps> = (props) => {
+  const { inspectionData, setInspectionData } = props;
+  const handleInspectorTabClick = useCallback(
+    (item: PivotItem | undefined) => {
```
Comment (hatpick) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx:49]: nit: item?: PivotItem

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+          setInspectionData({ ...inspectionData, mode: 'request' });
+        }
+        if (item.props.itemKey === 'networkRes') {
+          setInspectionData({ ...inspectionData, mode: 'response' });
+        }
+      }
+    },
+    [inspectionData, setInspectionData]
+  );
+  const renderHeader = (inspectionData: WebChatInspectionData) => {
```
Comment (hatpick) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx:61]: you can use a useCallback on this one too, saves some re-renders on the pivot

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        }
+        break;
+
+      case 'response':
+        if (inspectionData.item.trafficType === 'network') {
+          return inspectionData.item.response.payload;
+        }
+        break;
+
+      default:
```
Comment (hatpick) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx:99]: nit: remove

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+          left: true,
+          topRight: false,
+          bottomRight: false,
+          bottomLeft: false,
+          topLeft: false,
+        }}
+      >
+        {renderHeader(inspectionData)}
+        <JsonEditor
+          editorSettings={{ fadedWhenReadOnly: false }}
```
Comment (hatpick) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx:134]: pass editor settings from user settings for consistent font styles.

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebchatLogContent.tsx [hunk 10]
```diff
@@ -2,29 +2,97 @@
...
+  color: ${NeutralColors.gray130};
+`;
+
+const emptyStateMessage = css`
+  padding-left: 16px;
+`;
+
+const getActivityDirection = (activity) => {
+  if (activity?.recipient && activity.recipient.role === 'bot') {
+    return <span>{'->'}</span>;
```
Comment (hatpick) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebchatLogContent.tsx:82]: nit: dont need { and }

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebchatLogContent.tsx [hunk 11]
```diff
@@ -44,19 +112,123 @@ export const WebchatLogContent: React.FC<DebugPanelTabHeaderProps> = ({ isActive
     navigateToLatestEntryWhenActive(true);
   }, [currentLogItemCount]);
 
+  const onClickTraffic = useCallback((data: WebChatInspectionData) => {
+    if (currentProjectId) {
+      setWebChatInspectionData(currentProjectId, data);
+    }
+  }, []);
```
Comment (hatpick) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebchatLogContent.tsx:119]: currentProjectId

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebchatLogContent.tsx [hunk 12]
```diff
@@ -44,19 +112,123 @@ export const WebchatLogContent: React.FC<DebugPanelTabHeaderProps> = ({ isActive
...
+
+  const renderTimeStamp = (timestamp: string) => {
```
Comment (hatpick) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebchatLogContent.tsx:121]: nit: useCallback

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebchatLogContent.tsx [hunk 13]
```diff
@@ -44,19 +112,123 @@ export const WebchatLogContent: React.FC<DebugPanelTabHeaderProps> = ({ isActive
...
+          width: '100%',
+          display: 'flex',
+          overflow: 'auto',
+          flexDirection: 'column',
+          padding: '16px 0',
+          boxSizing: 'border-box',
+        }}
+        data-testid="Webchat-Logs-Container"
+      >
+        {displayedTraffic.length ? displayedTraffic : <span css={emptyStateMessage}>No Web Chat activity yet.</span>}
```
Comment (hatpick) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebchatLogContent.tsx:229]: formatMessage

File: Composer/packages/lib/code-editor/src/BaseEditor.tsx [hunk 14]
```diff
@@ -100,6 +100,7 @@ const mergeEditorSettings = (baseOptions: any, overrides: Partial<CodeEditorSett
     fontFamily: overrides?.fontSettings?.fontFamily,
     fontSize: overrides?.fontSettings?.fontSize,
     fontWeight: Number(overrides?.fontSettings?.fontWeight),
+    fadedWhenReadOnly: overrides?.fadedWhenReadOnly !== undefined ? overrides?.fadedWhenReadOnly : true,
```
Comment (hatpick) [inline Composer/packages/lib/code-editor/src/BaseEditor.tsx:103]: second ```overrides?.fadedWhenReadOnly``` can be ```overrides.fadedWhenReadOnly```

File: Composer/packages/client/src/components/WebChat/WebChatPanel.tsx (hunk 2 above)
Comment (tonyanziano) [reply Composer/packages/client/src/components/WebChat/WebChatPanel.tsx:82]: this was the behavior of the component before I added to it, so I decided to leave it as is

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx (hunk 4 above)
Comment (tonyanziano) [reply Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx]: I'm confused by this comment. This method is mutating state. Could you explain the scenarios in which you would use `onUpdateData` vs `setData` ?

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebchatLogContent.tsx (hunk 10 above)
Comment (tonyanziano) [reply Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebchatLogContent.tsx]: it's either `{'->'}` or `-&gt;` -- I think the former is more readable

Comment (beyackle) [reply Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebchatLogContent.tsx]: What if we just use &rarr; and get an actual right-arrow?

File: Composer/packages/client/src/components/WebChat/utils/conversationService.ts [hunk 15]
```diff
@@ -277,7 +276,7 @@ export class ConversationService {
       });
     } catch (ex) {
       const response: AxiosResponse = ex.response;
-      const err: DirectLineLog = {
+      const err = {
```
Comment (srinaath) [inline Composer/packages/client/src/components/WebChat/utils/conversationService.ts:279]: Isn't err still of the type DirectlineLog

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebChatLog/LogItemHelpers.tsx [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+  color: ${SharedColors.green20};
+  padding-right: 6px;
+`;
+
+const timestampBracket = css`
+  color: ${NeutralColors.gray130};
+`;
+
+export const renderActivityArrow = (activity) => {
+  if (activity?.recipient && activity.recipient.role === 'bot') {
```
Comment (srinaath) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebChatLog/LogItemHelpers.tsx:18]: activity?.recipient?.role instead &&?

File: Composer/packages/client/src/components/WebChat/utils/conversationService.ts (hunk 15 above)
Comment (tonyanziano) [reply Composer/packages/client/src/components/WebChat/utils/conversationService.ts:279]: I got rid of the `DirectlineLog` type

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebChatLog/WebChatActivityLogItem.tsx [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+import { renderTimeStamp } from './LogItemHelpers';
+import { clickableSegment, emphasizedText, hoverItem, logItem } from './logItemStyles';
+
+const clickable = css`
+  cursor: pointer;
+`;
+
+const renderActivityArrow = (activity) => {
+  if (activity?.recipient && activity.recipient.role === 'bot') {
```
Comment (srinaath) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebChatLog/WebChatActivityLogItem.tsx:19]: nit: activity?.recipient?.role.

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebChatLog/WebChatInspectorPane.tsx [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+
+type WebChatInspectorPaneProps = {
+  inspectionData?: WebChatInspectionData;
+  setInspectionData: (data: WebChatInspectionData) => void;
+};
+
+const getUserFriendlyActivityType = (type: ActivityType | undefined) => {
+  switch (type) {
+    case 'conversationUpdate':
+      return 'Conversation update';
```
Comment (srinaath) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebChatLog/WebChatInspectorPane.tsx:32]: nit: Conversation Update?

Comment (tonyanziano) [reply Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebChatLog/WebChatInspectorPane.tsx:32]: this capitalization is represented in the designs

File: Composer/packages/types/src/server.ts [hunk 19]
```diff
@@ -38,16 +38,54 @@ export type ISettingManager = {
...
+  response: NetworkTrafficResponse;
   timestamp: string;
-  route?: string;
-  logType: DirectLineLogType;
-}
+  trafficType: 'network';
+};
+
+export type ConversationActivityTraffic = {
+  activities: any[];
```
Comment (srinaath) [inline Composer/packages/types/src/server.ts:72]: We have botframework-types package. Can we include Activity type from there?

Comment (tonyanziano) [reply Composer/packages/types/src/server.ts:72]: Good call, let me try that

File: Composer/packages/client/src/components/WebChat/WebChatPanel.tsx (hunk 3 above)
Comment (tonyanziano) [reply Composer/packages/client/src/components/WebChat/WebChatPanel.tsx:97]: This is just on the server side, on the client side it gets transformed into a different format respective of the locale. I think I'm going to end up using `Intl.DateTimeFormat().format()` on the client though ([reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat))

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx (hunk 4 above)
Comment (hatpick) [reply Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx]: updates the state in the parent component, but for this component, it's a callback.
callbacks are named like OnSomethingHappened
state mutators are named like setSomething

File: Composer/packages/client/src/components/WebChat/WebChatPanel.tsx (hunk 3 above)
Comment (tonyanziano) [reply Composer/packages/client/src/components/WebChat/WebChatPanel.tsx:97]: Fixed

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx (hunk 4 above)
Comment (tonyanziano) [reply Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebchatLog/WebChatInspectorPane.tsx]: Got it, thanks :)

**Review by srinaath — APPROVED** [review]

**Review by hatpick — APPROVED** [review]

**Pull request merged**